src/main/java/bio/terra/stairway/Database.java

```diff
@@ -0,0 +1,460 @@
+
+        // Compute the table names
+        String namePrefix;
+        if (schemaName == null) {
+            namePrefix = "";
+        } else {
+            namePrefix = schemaName + '.';
+        }
+        if (nameStem != null) {
+            namePrefix = namePrefix + nameStem + '_';
+        }
+        this.flightVersionTableName = namePrefix + FLIGHT_VERSION_TABLE;
+        this.flightTableName = namePrefix + FLIGHT_TABLE;
+        this.flightLogTableName = namePrefix + FLIGHT_LOG_TABLE;
+
+        // Configure the database
+        if (forceCleanStart) {
+            startClean();
+        } else {
+            startDirty();
+        }
+    }
+
+    private boolean isDatabaseDisabled() {
+        return (dataSource == null);
+    }
+
+    /**
+     * If tables exist, drop them and re-create them with the version in this code.
+     */
+    private void startClean() {
+        try (Connection connection = dataSource.getConnection();
+             Statement statement = connection.createStatement()) {
+
+            statement.executeUpdate(
+                    "DROP TABLE IF EXISTS " +
+                            flightVersionTableName + "," +
+                            flightTableName + "," +
+                            flightLogTableName);
+
+        } catch (SQLException ex) {
+            throw new DatabaseSetupException("Failed to create database tables", ex);
+        }
+
+        // Create the schema from scratch
+        create();
+    }
+
+    /**
+     * If tables exist, read the schema version. Throw if the version is incompatible.
+     * If tables do not exist, create them.
+     * NOTE: this method does not (and cannot) initiate recovery. That is a separate
+     * operation driven from the Staircase class.
+     */
+    private void startDirty() {
+        try (Connection connection = dataSource.getConnection();
+             Statement statement = connection.createStatement()) {
+
+            schemaExists = versionTableExists(statement);
+            if (!schemaExists) {
+                create();
+                return;
+            }
+
+            // Schema exists. Pull out the schema version.
+            // TODO: make this more subtle. For now if the version doesn't exactly
+            // match we barf.
+            schemaVersion = readDatabaseSchemaVersion(statement);
+            if (!StringUtils.equals(schemaVersion, FLIGHT_SCHEMA_VERSION)) {
+                throw new DatabaseSetupException("Database schema version " + schemaVersion +
+                        "is incompatible with software schema version " + FLIGHT_SCHEMA_VERSION);
+            }
+
+        } catch (SQLException ex) {
+            throw new DatabaseSetupException("Failed to get schema version state", ex);
+        }
+    }
+
+    // Database accessors - package scoped so unit tests can use them
+    boolean versionTableExists(Statement statement) throws SQLException {
+        String query = "SELECT COUNT(*) AS version_table_count" +
```

Comment:
I'd prefer if we could avoid constructing sql like this. I know we own the inputs that go in here for now, but it is a good habit to get into. I know how the recommended way to do this in Python with psycopg2 is to use `%(named)s` tokens in the string and then pass in an object as a parameter to execute like `{'named': 'foo'}` to be safely interpolated.

It looks like prepared statements are the route to go with Java, can we try this?

Reply:
I have a TODO at the top of the database module saying I should do this. I figured that out late in the game. If it is OK, I would like to leave things as is for this PR and add a follow on story/PR to revamp the SQL construction.

Reply:
Sounds good. I made a story to track.

Reply:
To be done in follow-on PR